webnn/resources/utils.js

```diff
@@ -1,61 +1,350 @@
 'use strict';
 
-const ExecutionArray = ['async', 'sync'];
+const ExecutionArray = ['sync', 'async'];
 
 // https://webmachinelearning.github.io/webnn/#enumdef-mldevicetype
 const DeviceTypeArray = ['cpu', 'gpu'];
 
+// https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+const TypedArrayDict = {
+  float32: Float32Array,
+  int32: Int32Array,
+  uint32: Uint32Array,
+  int8: Int8Array,
+  uint8: Uint8Array,
+};
+
+const sizeOfShape = (array) => {
+  return array.reduce((accumulator, currentValue) => accumulator * currentValue, 1);
+};
+
+/**
+ * Get JSON resources from specified test resources file.
+ * @param {String} file - A test resources file path
+ * @returns {Object} Test resources
+ */
+const loadResources = (file) => {
+  const loadJSON = () => {
+    let xmlhttp = new XMLHttpRequest();
+    xmlhttp.open("GET", file, false);
+    xmlhttp.overrideMimeType("application/json");
+    xmlhttp.send();
+    if (xmlhttp.status == 200 && xmlhttp.readyState == 4) {
+      return xmlhttp.responseText;
+    } else {
+      throw new Error(`Failed to load ${file}`);
+    }
+  };
+
+  const json = loadJSON();
+  return JSON.parse(json.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m));
+};
+
+/**
+ * Get exptected data from given resources with output name.
+ * @param {Array} resources - An array of expected resources
+ * @param {String} outputName - An output name
+ * @returns {Number[]} An expected data array
+ */
+const getExpectedData = (resources, outputName) => {
+  let data;
+  for (let subResources of resources) {
+    if (subResources.name === outputName) {
+      data = subResources.data;
+      break;
+    }
+  }
+  if (data === undefined) {
+    throw new Error(`Failed to get expected data sources by ${outputName}`);
+  }
+  return data;
+};
+
+/**
+ * Get ULP tolerance of softmax operation.
+ * @param {Object} resources - Resources used for building a graph
+ * @returns {Number} A tolerance number
+ */
+const getSoftmaxPrecisionTolerance = (resources) => {
+  // Compute the softmax values of the 2-D input tensor along axis 1.
+  const inputShape = resources.inputs[Object.keys(resources.inputs)[0]].shape;
+  const tolerance = inputShape[1] * 2 + 3;
```

Comment:
`IEPOE*3+3 ULP` Sorry for the typo. I updated it here: https://github.com/webmachinelearning/webnn/issues/265. The subtraction, exponentiation, and addition all contribute to loss in `expₑ(a - reducemax(A, axes)) / reducesum(expₑ(A - reducemax(A, axes)), axis)`.

Reply:
I've updated, thanks @fdwr